http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -40,24 +44,87 @@ class ClusterHelper(
+      case None => IO.unit
+      case Some(imageProject) =>
+        IO.fromFuture(IO(dbRef.inTransaction {
+            _.clusterQuery.countActiveByProject(googleProject)
+          }))
+          .flatMap { count =>
+            // Note: Don't remove the account if there are existing active clusters in the same project,
+            // because it could potentially break other clusters. We only check this for the 'remove' case.
+            if (count > 0 && !createCluster) {
+              IO.unit
+            } else {
+              val projectNumberOptIO = IO.fromFuture(IO(googleComputeDAO.getProjectNumber(googleProject)))
+              val clusterIamSetupExceptionIO = IO.raiseError[Long](ClusterIamSetupException(imageProject))
+              for {
+                projectNumber <- projectNumberOptIO.flatMap(_.fold(clusterIamSetupExceptionIO)(IO.pure))
+                // Note that the Dataproc service account is used to retrieve the image, and not the user's
+                // pet service account. There is one Dataproc service account per Google project. For more details:
+                // https://cloud.google.com/dataproc/docs/concepts/iam/iam#service_accounts
+                dataprocServiceAccountEmail = WorkbenchEmail(
+                  s"service-${projectNumber}@dataproc-accounts.iam.gserviceaccount.com"
+                )
+                _ <- updateGroupMembership(googleGroupsConfig.dataprocImageProjectGroupEmail,
+                                           dataprocServiceAccountEmail,
+                                           createCluster)
+              } yield ()
+            }
+          }
+    }
+
+  /**
+   * Add the Dataproc Worker role in the user's project to the cluster service account, if present.
+   * This is needed to be able to spin up Dataproc clusters using a custom service account.
+   * If the Google Compute default service account is being used, this is not necessary.
+   */
   private def updateClusterIamRoles(googleProject: GoogleProject,
                                     serviceAccountInfo: ServiceAccountInfo,
                                     createCluster: Boolean): IO[Unit] = {
-    val retryIam: (GoogleProject, WorkbenchEmail, Set[String]) => IO[Boolean] = (project, email, roles) =>
-      IO.fromFuture[Boolean](IO(retryExponentially(when409, s"IAM policy change failed for Google project '$project'") {
+    val retryIam: (GoogleProject, WorkbenchEmail, Set[String]) => IO[Unit] = (project, email, roles) =>
+      IO.fromFuture[Unit](IO(retryExponentially(when409, s"IAM policy change failed for Google project '$project'") {
         () =>
-          if (createCluster) {
+          val iamUpdate = if (createCluster) {
             googleIamDAO.addIamRoles(project, email, MemberType.ServiceAccount, roles)
           } else {
             googleIamDAO.removeIamRoles(project, email, MemberType.ServiceAccount, roles)
           }
+          iamUpdate map { _ =>
+            ()
+          }
```

Comment:
This ugliness is for the method to yield an `IO[Unit]` (from `IO[Boolean]` which `add/removeIamRoles()` returns) 🙈 This type difference existed in the dev version as well but there was another IO construction there, which somehow makes the types fit.